This notebook contains the hyperparameter tuning for the different models.

# DATA LOADING


In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('data_up.csv')

In [3]:
data.head(5)

SEQN  Gender  Age  Annual-Family-Income  Ratio-Family-Income-Poverty  \
0     2       1   77                     8                         5.00   
1     5       1   49                    11                         5.00   
2    12       1   37                    11                         4.93   
3    13       1   70                     3                         1.07   
4    14       1   81                     5                         2.67   

   X60-sec-pulse  Systolic  Diastolic  Weight  Height  ...  Total-Cholesterol  \
0             68        98         56    75.4   174.0  ...               5.56   
1             66       122         83    92.5   178.3  ...               7.21   
2             64       174         99    99.2   180.0  ...               4.03   
3            102       130         66    63.6   157.7  ...               8.12   
4             72       136         61    75.5   166.2  ...               4.50   

    HDL  Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
0  1.39              4.7              3              3                 1   
1  1.08              5.5              1              1                 1   
2  0.98              5.2              2              1                 1   
3  1.28              7.6              3              3                 1   
4  1.04              5.8              1              1                 1   

   Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  CoronaryHeartDisease  
0         2                   2                 2                     0  
1         2                   2                 2                     0  
2         2                   1                 1                     0  
3         1                   1                 2                     0  
4         2                   2                 2                     0  

[5 rows x 51 columns]

In [4]:
data.columns

Index(['SEQN', 'Gender', 'Age', 'Annual-Family-Income',
       'Ratio-Family-Income-Poverty', 'X60-sec-pulse', 'Systolic', 'Diastolic',
       'Weight', 'Height', 'Body-Mass-Index', 'White-Blood-Cells',
       'Lymphocyte', 'Monocyte', 'Eosinophils', 'Basophils', 'Red-Blood-Cells',
       'Hemoglobin', 'Mean-Cell-Vol', 'Mean-Cell-Hgb-Conc.',
       'Mean-cell-Hemoglobin', 'Platelet-count', 'Mean-Platelet-Vol',
       'Segmented-Neutrophils', 'Hematocrit', 'Red-Cell-Distribution-Width',
       'Albumin', 'ALP', 'AST', 'ALT', 'Cholesterol', 'Creatinine', 'Glucose',
       'GGT', 'Iron', 'LDH', 'Phosphorus', 'Bilirubin', 'Protein', 'Uric.Acid',
       'Triglycerides', 'Total-Cholesterol', 'HDL', 'Glycohemoglobin',
       'Vigorous-work', 'Moderate-work', 'Health-Insurance', 'Diabetes',
       'Blood-Rel-Diabetes', 'Blood-Rel-Stroke', 'CoronaryHeartDisease'],
      dtype='object')

In [5]:
data.shape

(37079, 51)

In [6]:
data.isnull().sum()

SEQN                           0
Gender                         0
Age                            0
Annual-Family-Income           0
Ratio-Family-Income-Poverty    0
X60-sec-pulse                  0
Systolic                       0
Diastolic                      0
Weight                         0
Height                         0
Body-Mass-Index                0
White-Blood-Cells              0
Lymphocyte                     0
Monocyte                       0
Eosinophils                    0
Basophils                      0
Red-Blood-Cells                0
Hemoglobin                     0
Mean-Cell-Vol                  0
Mean-Cell-Hgb-Conc.            0
Mean-cell-Hemoglobin           0
Platelet-count                 0
Mean-Platelet-Vol              0
Segmented-Neutrophils          0
Hematocrit                     0
Red-Cell-Distribution-Width    0
Albumin                        0
ALP                            0
AST                            0
ALT                            0
Cholesterol                    0
Creatinine                     0
Glucose                        0
GGT                            0
Iron                           0
LDH                            0
Phosphorus                     0
Bilirubin                      0
Protein                        0
Uric.Acid                      0
Triglycerides                  0
Total-Cholesterol              0
HDL                            0
Glycohemoglobin                0
Vigorous-work                  0
Moderate-work                  0
Health-Insurance               0
Diabetes                       0
Blood-Rel-Diabetes             0
Blood-Rel-Stroke               0
CoronaryHeartDisease           0
dtype: int64

Conclusion : Data is clean and all the data is numerical
CoronaryHeartDisease is target class. This is a classification problem predicting the heart disease.
SEQN is just representing the row number hence we need not take in the prediction columns.
Here, we have 49 features and 1 target column.
We have originally 37079 rows with no cell being empty.

# DATA ANALYSIS


In [7]:
data.describe()

SEQN        Gender           Age  Annual-Family-Income  \
count  37079.000000  37079.000000  37079.000000          37079.000000   
mean   48901.041236      1.513282     48.943661              7.358208   
std    26753.636441      0.499830     18.010440              3.994083   
min        2.000000      1.000000     20.000000              1.000000   
25%    26120.500000      1.000000     33.000000              4.000000   
50%    50065.000000      2.000000     48.000000              7.000000   
75%    71173.500000      2.000000     63.000000             10.000000   
max    93702.000000      2.000000     85.000000             15.000000   

       Ratio-Family-Income-Poverty  X60-sec-pulse      Systolic     Diastolic  \
count                 37079.000000   37079.000000  37079.000000  37079.000000   
mean                      2.559026      72.579250    124.090078     69.919253   
std                       1.624789      12.242108     19.254741     13.575804   
min                       0.000000      32.000000      0.000000      0.000000   
25%                       1.140000      64.000000    111.000000     62.000000   
50%                       2.180000      72.000000    121.000000     70.000000   
75%                       4.130000      80.000000    134.000000     78.000000   
max                       5.000000     224.000000    270.000000    132.000000   

             Weight        Height  ...  Total-Cholesterol           HDL  \
count  37079.000000  37079.000000  ...       37079.000000  37079.000000   
mean      80.988276    167.389601  ...           5.081713      1.370344   
std       20.678734     10.122908  ...           1.072682      0.415985   
min       32.300000    129.700000  ...           1.530000      0.160000   
25%       66.500000    160.000000  ...           4.320000      1.070000   
50%       78.200000    167.100000  ...           5.020000      1.290000   
75%       92.100000    174.600000  ...           5.740000      1.600000   
max      371.000000    204.500000  ...          14.090000      5.840000   

       Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
count     37079.000000   37079.000000   37079.000000      37079.000000   
mean          5.676496       1.783840       1.598856          1.218587   
std           1.050223       0.448324       0.511199          0.461102   
min           2.000000       1.000000       1.000000          1.000000   
25%           5.200000       2.000000       1.000000          1.000000   
50%           5.400000       2.000000       2.000000          1.000000   
75%           5.800000       2.000000       2.000000          1.000000   
max          18.800000       3.000000       3.000000          9.000000   

           Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  \
count  37079.000000        37079.000000      37079.000000   
mean       1.907333            1.549502          1.796165   
std        0.349674            0.497550          0.402853   
min        1.000000            1.000000          1.000000   
25%        2.000000            1.000000          2.000000   
50%        2.000000            2.000000          2.000000   
75%        2.000000            2.000000          2.000000   
max        3.000000            2.000000          2.000000   

       CoronaryHeartDisease  
count          37079.000000  
mean               0.040670  
std                0.197527  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000  

[8 rows x 51 columns]

In [8]:
data['Pulse Pressure'] = data['Diastolic'] - data['Systolic']

In [9]:
data['BMI'] = data['Weight'] / (data['Height'] ** 2)

In [10]:
data = data.drop(columns=['Diastolic', 'Systolic', 'Height', 'Weight'])

In [11]:
correlation_matrix = data.corr()

In [12]:
correlation_matrix

SEQN    Gender       Age  \
SEQN                         1.000000 -0.001776  0.001023   
Gender                      -0.001776  1.000000 -0.031112   
Age                          0.001023 -0.031112  1.000000   
Annual-Family-Income         0.178423 -0.044304 -0.042854   
Ratio-Family-Income-Poverty -0.044699 -0.051150  0.051216   
X60-sec-pulse                0.011624  0.154636 -0.149767   
Body-Mass-Index              0.059610  0.059787  0.029349   
White-Blood-Cells           -0.006061  0.041347 -0.073692   
Lymphocyte                   0.050807  0.043523 -0.097218   
Monocyte                     0.006424 -0.198896  0.153205   
Eosinophils                  0.009960 -0.124137  0.073331   
Basophils                    0.096240  0.012746  0.044337   
Red-Blood-Cells             -0.020411 -0.481395 -0.157705   
Hemoglobin                  -0.090237 -0.560418 -0.064956   
Mean-Cell-Vol               -0.082581 -0.106334  0.192301   
Mean-Cell-Hgb-Conc.         -0.097088 -0.114206  0.136862   
Mean-cell-Hemoglobin        -0.080280 -0.052393 -0.059845   
Platelet-count              -0.179958  0.221401 -0.140836   
Mean-Platelet-Vol            0.044662  0.030218 -0.021834   
Segmented-Neutrophils       -0.052009  0.035817  0.031735   
Hematocrit                  -0.074596 -0.563799 -0.050066   
Red-Cell-Distribution-Width  0.262741  0.090150  0.157143   
Albumin                     -0.002825 -0.294608 -0.132660   
ALP                         -0.119024 -0.006748  0.112062   
AST                          0.015769 -0.104826  0.018363   
ALT                         -0.009044 -0.156318 -0.054341   
Cholesterol                 -0.090179  0.078611  0.100606   
Creatinine                   0.038944 -0.282505  0.197243   
Glucose                      0.074342 -0.061140  0.223688   
GGT                         -0.023706 -0.118801  0.044975   
Iron                        -0.049586 -0.212976 -0.045462   
LDH                         -0.137315 -0.023433  0.209670   
Phosphorus                   0.068226  0.146123 -0.078160   
Bilirubin                   -0.074459 -0.268788  0.031654   
Protein                     -0.177187 -0.108174 -0.082856   
Uric.Acid                    0.022743 -0.432240  0.163410   
Triglycerides                0.030145 -0.114124  0.072948   
Total-Cholesterol           -0.122045  0.074376  0.109633   
HDL                          0.023288  0.308572  0.050450   
Glycohemoglobin              0.085223 -0.042684  0.283285   
Vigorous-work                0.054571  0.151403  0.207679   
Moderate-work                0.056670  0.058853  0.117950   
Health-Insurance            -0.001002 -0.049752 -0.243152   
Diabetes                    -0.032711  0.014295 -0.190432   
Blood-Rel-Diabetes           0.054461 -0.054130 -0.000434   
Blood-Rel-Stroke             0.176478 -0.043613  0.024753   
CoronaryHeartDisease        -0.004985 -0.078408  0.222649   
Pulse Pressure               0.012633  0.005023 -0.479900   
BMI                          0.059638  0.059808  0.029341   

                             Annual-Family-Income  \
SEQN                                     0.178423   
Gender                                  -0.044304   
Age                                     -0.042854   
Annual-Family-Income                     1.000000   
Ratio-Family-Income-Poverty              0.871092   
X60-sec-pulse                           -0.056541   
Body-Mass-Index                         -0.025350   
White-Blood-Cells                       -0.078530   
Lymphocyte                               0.017299   
Monocyte                                 0.014725   
Eosinophils                             -0.005209   
Basophils                                0.017522   
Red-Blood-Cells                          0.032173   
Hemoglobin                               0.031878   
Mean-Cell-Vol                           -0.013885   
Mean-Cell-Hgb-Conc.                      0.000438   
Mean-cell-Hemoglobin                     0.038898   
Platelet-count              

# Feature Selection


**ALGORITHM : BMPA**

In [13]:
!ls -l


total 8184
-rw-r--r-- 1 root root 8374838 Oct 13 16:05 data_up.csv
drwxr-xr-x 1 root root    4096 Oct  9 13:36 sample_data


In [14]:
!pip install mealpy

In [15]:
import mealpy
import pkgutil
mods = [name for _, name, _ in pkgutil.walk_packages(mealpy.__path__, mealpy.__name__ + ".")]

print([m for m in mods if "mpa" in m.lower() or "bmpa" in m.lower()])


['mealpy.swarm_based.MPA']


In [16]:
!pip install --upgrade git+https://github.com/thieu1995/mealpy.git scikit-learn pandas numpy -q

  Preparing metadata (setup.py) ... done


In [17]:
import pandas as pd, numpy as np
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from mealpy.swarm_based.MPA import OriginalMPA
from mealpy.utils.space import BinaryVar

Load dataset


In [18]:
df = pd.read_csv('data_up.csv')

In [19]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])

In [20]:
y = df['CoronaryHeartDisease']

Here we are defining fitness function by (minimize 1- accuracy)
And also we Train/Test split



In [21]:
def fitness(sol):
    mask = sol > 0.5
    if not mask.any():
        return 1
    Xs = X.loc[:, mask]

    Xtr, Xte, ytr, yte = train_test_split(Xs, y, test_size=0.2, random_state=42)
    clf = RandomForestClassifier(random_state=42).fit(Xtr, ytr)

    return 1 - accuracy_score(yte, clf.predict(Xte))

In [22]:
bounds = [BinaryVar() for _ in range(X.shape[1])]

Defining Optimization Problem




In [23]:
problem = {
    "obj_func": fitness,
    "bounds": bounds,
    "minmax": "min",
}

Initializing Model and Running Binary MPA

In [24]:
model = OriginalMPA(epoch=5, pop_size=20)
result = model.solve(problem)

INFO:mealpy.swarm_based.MPA.OriginalMPA:OriginalMPA(epoch=5, pop_size=20)
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 1, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 148.30503 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 2, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 160.68031 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 3, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 162.87467 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 4, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 165.66727 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 5, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 173.85492 seconds


In [25]:
best_position = result.solution

if hasattr(result, "solution_fitness"):
    best_fitness = result.solution_fitness
elif hasattr(result, "fitness_history"):
    best_fitness = result.fitness_history[-1]
else:
    best_fitness = None
    print("Could not find fitness attribute in result. Inspect result.__dict__")

Could not find fitness attribute in result. Inspect result.__dict__


Selected Features


In [26]:
selected_features = np.where(best_position > 0.05)[0]
X_selected = X.iloc[:, selected_features] if selected_features.size > 0 else pd.DataFrame()

Calculating the Actual Accuracy and printing the results

In [27]:
if X_selected.shape[1] > 0:
    clf_final = RandomForestClassifier(random_state=42)
    accuracy = np.mean(cross_val_score(clf_final, X_selected, y, cv=5))
else:
    accuracy = 0



In [28]:
print("Best Fitness (1 - Accuracy):", best_fitness)


Best Fitness (1 - Accuracy): None


In [29]:
print("Selected Features:", selected_features)


Selected Features: [ 0  1  4  6  7  8 10 11 12 13 14 15 17 20 21 22 24 26 27 30 33 34 36 39
 40 46 48]


In [30]:
print("Total Features Selected:", len(selected_features))


Total Features Selected: 27


In [31]:
print("Accuracy of Selected Features:", accuracy)

Accuracy of Selected Features: 0.9593300810843644


# MODEL SELECTION


In [32]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

models = {
    'RandomForestClassifier' : RandomForestClassifier(random_state=42),
    'LogisticRegression' : LogisticRegression(random_state=42),
}

In [33]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])
y = df['CoronaryHeartDisease']
filtered_X = X.iloc[:, selected_features]

In [34]:
filtered_X

Gender  Age  X60-sec-pulse  Diastolic  Weight  Height  \
0           1   77             68         56    75.4   174.0   
1           1   49             66         83    92.5   178.3   
2           1   37             64         99    99.2   180.0   
3           1   70            102         66    63.6   157.7   
4           1   81             72         61    75.5   166.2   
...       ...  ...            ...        ...     ...     ...   
37074       1   25             92         76    39.2   136.5   
37075       2   76             78         46    59.1   165.8   
37076       2   80             74         58    71.7   152.2   
37077       1   35             76         66    78.2   173.3   
37078       2   24             80         68    58.3   165.0   

       White-Blood-Cells  Lymphocyte  Monocyte  Eosinophils  ...  ALP  AST  \
0                    7.6        21.1       7.1          4.4  ...   62   19   
1                    5.9        37.8       6.2          3.4  ...   63   22   
2                   10.2        23.7       9.0          3.2  ...   63   17   
3                   11.6        13.1       3.8          0.4  ...  103   24   
4                    9.1        29.8       5.6          1.7  ...  110   23   
...                  ...         ...       ...          ...  ...  ...  ...   
37074                7.1        24.2       8.9          0.9  ...   67   34   
37075                6.4        21.0      12.7          3.9  ...   50   22   
37076                4.7        45.9      12.5          3.8  ...   54   25   
37077                7.6        26.4       9.2          2.0  ...  140   73   
37078                9.7        24.6       5.5          0.5  ...   35   24   

       Creatinine   Iron  LDH  Bilirubin  Triglycerides  Total-Cholesterol  \
0           61.90  11.28  140      12.00          1.298               5.56   
1           70.70  24.54  133       8.60          3.850               7.21   
2           88.40  11.28  131       6.80          1.581               4.03   
3           61.90  12.18  181       8.60          3.635               8.12   
4           88.40  11.82  150      10.30          0.756               4.50   
...           ...    ...  ...        ...            ...                ...   
37074       56.58  13.10  144      11.97          1.264               4.14   
37075       97.24  15.00  124      18.81          0.948               3.62   
37076       81.33   8.40  120       5.13          1.095               6.62   
37077       82.21   9.00  136       3.42          0.937               3.72   
37078       58.34  27.90   93       8.55          0.892               4.14   

       Diabetes  Blood-Rel-Stroke  
0             2                 2  
1             2                 2  
2             2                 1  
3             1                 2  
4             2                 2  
...         ...               ...  
37074         2                 2  
37075         2                 2  
37076         2                 2  
37077         2                 1  
37078         2                 2  

[37079 rows x 27 columns]

In [35]:
for name, model in models.items():
    # Use the filtered_X data for model selection
    X_train, X_test, y_train, y_test = train_test_split(filtered_X, y, test_size=0.2, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"{name} Accuracy: {accuracy}")

RandomForestClassifier Accuracy: 0.9584681769147788
LogisticRegression Accuracy: 0.9572545846817692


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# Hyperparameter Tuning
Tune the hyperparameters of the models using GridSearchCV.

## Define parameter grids

### Subtask:
Define parameter grids for each model to search during tuning.


**Reasoning**:
Defining parameter grids for each model as requested in the instructions.



In [36]:
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 6],
    'min_samples_split': [2, 5],
  }

param_grid_lr = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'penalty': ['l2'],
    'solver': ['lbfgs']
}

## Set up gridsearchcv

### Subtask:
Use GridSearchCV for each model with appropriate cross-validation.


**Reasoning**:
Import GridSearchCV and create GridSearchCV objects for each model with the defined parameter grids and cross-validation.



In [37]:
from sklearn.model_selection import GridSearchCV

tuned_models = {}
param_grids = {
    'RandomForestClassifier': param_grid_rf,
    'LogisticRegression': param_grid_lr
}

for name, model in models.items():
    if name in param_grids:
        grid_search = GridSearchCV(estimator=model, param_grid=param_grids[name], cv=5, scoring='accuracy')
        tuned_models[name] = grid_search


## Perform tuning

### Subtask:
Fit the GridSearchCV objects to the selected data.


**Reasoning**:
Fit each GridSearchCV object to the selected features and target variable.



In [38]:
subset_size = 0.3
X_subset, _, y_subset, _ = train_test_split(filtered_X, y, test_size=1 - subset_size, random_state=42)

for name, grid_search in tuned_models.items():
    print(f"Tuning {name}...")
    grid_search.fit(X_subset, y_subset)
    print(f"Finished tuning {name}.")

Tuning RandomForestClassifier...
Finished tuning RandomForestClassifier.
Tuning LogisticRegression...


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to

Finished tuning LogisticRegression.


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [39]:
results = {}
for name, grid_search in tuned_models.items():
    best_model = grid_search.best_estimator_
    X_train, X_test, y_train, y_test = train_test_split(filtered_X, y, test_size=0.2, random_state=42)
    best_model.fit(X_train, y_train)
    y_pred = best_model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    results[name] = accuracy
    print(f"Tuned {name} Accuracy: {accuracy}")

Tuned RandomForestClassifier Accuracy: 0.9583333333333334
Tuned LogisticRegression Accuracy: 0.9571197411003236


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
